src/NServiceBus.Core.Tests/Routing/RoutingToDispatchConnectorTests.cs

```diff
@@ -0,0 +1,44 @@
+﻿namespace NServiceBus.Core.Tests.Routing
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Extensibility;
+    using NServiceBus.Pipeline;
+    using NServiceBus.Routing;
+    using NServiceBus.Transports;
+    using NUnit.Framework;
+    using ObjectBuilder;
+
+    [TestFixture]
+    public class RoutingToDispatchConnectorTests
+    {
+        [Test]
+        public async Task It_preserves_headers_generated_by_custom_routing_strategy()
+        {
+            var behavior = new RoutingToDispatchConnector();
+            var message = new OutgoingMessage("ID", new Dictionary<string, string>(), new byte[0]);
+            await behavior.Invoke(new RoutingContext(message,
+                new CustomRoutingStrategy(), new FakeContext()), context =>
+                {
+                    Assert.IsTrue(context.Operations.First().Message.Headers.ContainsKey("CustomHeader"));
```

Comment:
I think we should have the assertions at the end of the test and not within a callback unless necessary.